man/perfect.1
.TH PERFECT 1 "7 Aug 1995" "perfect"
.UC
.SH NAME
perfect \- create perfect resources
.SH SYNOPSIS
.B perfect
[
.BI \-e\ configfile
]
[
.BI \-s\ sectfile
]
.br
.SH DESCRIPTION
While most deities prefer to use
.I fairland
to generate their resources, this program is included as an alternative for
creating the resources for a game.  It sets all resources to 100.
.SH OPTIONS
.TP
.BI \-e\ configfile
causes
.I perfect
to use the game configuration specified in 
.I configfile
(by default, the file "econfig" in the data directory is used)
.TP
.BI \-s\ sectfile
give resources to the sectors in 
.I sectfile
(default is the "sector" file in the data directory)


doc/CHANGES
This file contains a list of patches/upgrades made to the server for a specific
version.  It also contains the name of the author of that particular code.  It
may not be complete.

*version 2.0 beta 1 release 1*
Fixed age.c so reserves degrade normally instead of growing! - Chad Zabel
Fixed arch.c - replaced wrapp with what was supposed to be there - Ken Huisman
Patched bug in atta.c - a temporary patch that could be done better
( and will be later ) - Ken H.
patched bug in lib/commands/mail.c - Dave Pare
added patches to make code compile under linux - Ken H. based on linux chain 3
patches by someone else whose name I forget.  Note at this time the code
*compiles* under linux, but there are still bugs - the server seems to
stop taking socket connections or something after someone logs out of the
game.
fixed a null pointer bug in lib/subs/bigmap.c - Chad Zabel

*version 2.0 beta 1 release 2*
Patch for exec bug - Chad Zabel

*version 2.0 beta 1 release 3*
Patch for zepplin bug - Jim Ortlieb
Added unixware(svr4.2) specific stuff to arch.c - Doug Hay
Changes to h/lwp.h and lib/lwp/lwp.c for thread stack checker - Doug Hay

*version 2.0 beta 1 release 4*
Modified fairland.c to 1.6.2 and made it so translate_continents is only
called when BLITZ is defined  - Ken Huisman, Ken Stevens
Modifications to lwp.h and arch.c to fix linux threads - Scott Ferguson

*version 2.0 beta 1 release 5*
Changed version number to 2.0.0 - Ken Huisman
fixed unterminated path bug in lib/common/path.c - Scott Ferguson
fixed first part of bug described in wish list item c6 ( false resupply ) -
Scott Ferguson
fixed part of bug described in wish list item t3 ( you have 666 new telegrams
bug ) - the bug that tells you you have 1 new telex after just reading it is
not fixed - Scott Ferguson

*version 2.0 beta 1 release 6*
Three letter abbreviations for all ships in ship.c - Chad Zabel
the following by Kevin Klemmick:
/lib/commands/rea.c                     // fixed telegram's problem reading '%'
/lib/commands/mail.c                    // ditto for mail and wmail
/lib/subs/aircombat.c                   // New air code and bug fixes
/lib/subs/mission.c                     // New air code and bug fixes
/lib/subs/pr.c                          // new prnf function an bug fixes
/lib/subs/paths.c                       // fixes the getpath() bug.
added NEWPAF and HIDDEN to options.h ( commented out for now until
testing is complete ) - Ken H.

*version 2.0 beta 1 release 7*
Fix for fix for exec bug ( lib/player/player.c ) - Chad Zabel
fixed minor bug in aircombat.c ( missing quote on line 613 ) - Ken H.

*version 2.0 beta 1 release 8*
lib/commands/mfir.c Bugfix... in quiet_bigdef, added in a check
for forts >60%.  They were getting cheated out of their bonus in the return
fire process. - Chad Zabel
AIX Port: files
lib/lwp/{GNUmakefile,arch.c,lwp.c,lwpint.h} and                             
make/Systems/aix modified
 - All changes are surrounded by #ifdef AIX32. 
 - one change to lwp.c that will apply to everyone one error
   message changed so it says underflow instead of overflow in the stack
   checking code.
- Bill Canning

*version 2.0 beta 1 release 9*
You can now specify how many mil to board a ship with (wishlist item a1) so
you don't strand your ship.  This is implemented as another parameter
in the board command.  Documentation for board updated too!  - Chad Zabel

- Addition of the HIDDEN and NEWPAF Mods (Major addition). - Kevin Klemmick

*version 2.0 beta 1 release 10*
lib/player/player.c compilation under linux was broken by previous patches
this is fixed now - Ken H.

*version 2.0 beta 1 release 11*
fixed compilation under sunos (lib/lwp/lwp.c had a couple odd '\' characters
in it that caused the sun compiler to complain) - Ken H.

*version 2.0 beta 1 release 12*
added 3 character abbreviations for land unit names - Chad Zabel
fixed bug in "show sect build" command - Ken Huisman
added check in neweff command to see if sector is stopped - Chad Zabel
added test for war declaration before abm launch - Chad Zabel

*version 2.0 beta 1 release 13*
added thread debugging stuff - Dave Pare
changes to autonav code - Chad Zabel
fixes for minor bug in land unit names - Chad Zabel
fix for build.t in docs - Chad Zabel
simple fix for spy command - Chad Zabel
fix for telegrams - Julian Onions
fixup of makefile in util - Julian Onions

*version 2.0 beta 1 release 14*
fix bugs in lwp - Julian Onions
fix bug in edit that caused server crash - Julian Onions
fix bugs in attack command that allowed free resource creation - Chad Zabel
fix bug that made countries in sanctuary lose money - Chad Zabel

*version 2.0 beta 1 release 15*
io bug fixes and misc code changes - Dave Pare
removal of sanctuary print statements to reduce logfile sizes - Ken Huisman
DEC Alpha port - Dave Pare and Ken Stevens

*version 2.0 beta 1 release 16*
made compiling/configuring easier (see READ_THIS_FIRST) - Ken Stevens
fix bug in info.c that caused crash - Janjaap van Velthooven
fix for land units - Chad Zabel

*version 2.0 beta 1 release 17*
major improvements/cleanup of info pages and related files - Ken Stevens
changes to compile scripts - Ken Stevens

*version 2.0 beta 1 release 18*
fixed missile sub abbreviation in ship file - Ken Stevens
added "are you sure" in build command to prevent accidents - Ken Stevens
fixed production simulation report showing "e" sectors producing - Ken Stevens
fixed LandUnits subject not being listed in "info". - Ken Stevens
fixed so unload only works in your sector or ally's sector - Ken Stevens
fixed error in sweep when sector efficiency is printed - Ken Stevens
fixed telegrams reporting bombing damage - Ken Stevens
fixed cede so you can only cede to allies - Ken Stevens
fixed coastwatch to show ship names - Ken Stevens
build command gives informative error message if you are building
in the wrong type of sector or the sector is < 60% eff. - Ken Stevens
build command does not prompt for tech level if not given on command

src/util/perfect.c
/*
 *  Empire - A multi-player, client/server Internet based war game.
 *  Copyright (C) 1986-2000, Dave Pare, Jeff Bailey, Thomas Ruschak,
 *                           Ken Stevens, Steve McClure
 *
 *  This program is free software; you can redistribute it and/or modify
 *  it under the terms of the GNU General Public License as published by
 *  the Free Software Foundation; either version 2 of the License, or
 *  (at your option) any later version.
 *
 *  This program is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU General Public License for more details.
 *
 *  You should have received a copy of the GNU General Public License
 *  along with this program; if not, write to the Free Software
 *  Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
 *
 *  ---
 *
 *  See the "LEGAL", "LICENSE", "CREDITS" and "README" files for all the
 *  related information and legal notices. It is expected that any future
 *  projects/authors will amend these files as needed.
 *
 *  ---
 *
 *  perfect.c: Create a perfect world
 * 
 *  Known contributors to this file:
 *     James Anderson, 1986
 *     Dave Pare, 1986
 */

#if defined(aix) || defined(linux) || defined(solaris)
#include <unistd.h>
#endif /* aix or linux */

#include <stdio.h>
#include <fcntl.h>
#include "misc.h"
#include "var.h"
#include "sect.h"
#include "gamesdef.h"
#include "file.h"
#include "path.h"
#include "xy.h"
#include "prototypes.h"

#define XPLATES		(WORLD_X / 2)
#define YPLATES		WORLD_Y
#define YbyX		(YPLATES*XPLATES/8)

struct sctstr sects[YPLATES * XPLATES];
u_char s[YPLATES][XPLATES];
short c[YPLATES][XPLATES];
short e[YPLATES][XPLATES];
int size[YbyX];
int goldchance[YbyX];
int ironchance[YbyX];
int oilchance[YbyX];
int fertchance[YbyX];
int urchance[YbyX];

void makec();
void makeore();
int nearelev();

int
main(argc, argp)
int argc;
s_char **argp;
{
    register struct sctstr *sp;
    register int y;
    register int x;
    int sectf;
    int i;
    time_t now;
    s_char *sectfil = NULL;
    extern char *optarg;
    int opt;
    char *config_file = NULL;


    while ((opt = getopt(argc, argp, "e:s:")) != EOF) {
	switch (opt) {
	case 'e':
	    config_file = optarg;
	    break;
	case 's':
	    sectfil = optarg;
	    break;
	}
    }

    emp_config(config_file);

    if (sectfil == NULL)
	sectfil = empfile[EF_SECTOR].file;

    time(&now);
    srandom((unsigned int)now);
    sectf = open(sectfil, O_RDWR, 0);
    if (sectf < 0) {
	perror(sectfil);
	exit(1);
    }
    read(sectf, sects, sizeof(sects));
    sp = sects;
    for (y = 0; y < WORLD_Y; y++) {
	for (x = 0; x < WORLD_X; x++) {
	    if (((x ^ y) & 1))
		continue;
	    c[y][x / 2] = 0;
	    s[y][x / 2] = sp->sct_type;
	    e[y][x / 2] = sp->sct_elev;
	    sp++;
	}
    }
    for (y = 0; y < WORLD_Y; y++) {
	for (x = 0; x < WORLD_X; x++) {
	    if (((x ^ y) & 1))
		continue;
	    if (c[y][x / 2] != 0)
		continue;
	    makec(x, y, 0);
	}
    }
    printf("size  oil  gold fert iron\n");
    for (i = 1; i < YbyX && size[i]; i++) {
	oilchance[i] = 5 + roll(75) - min(size[i] / 15, 10);
/*		goldchance[i] = 3 + roll(45) -  */
	goldchance[i] = 3 + roll(55) -
	    (oilchance[i] + min(size[i] / 15, 10));
/*		urchance[i] = roll(40) - goldchance[i]; */
	urchance[i] = roll(30) - goldchance[i];
	fertchance[i] = roll(40) - (goldchance[i] + oilchance[i]);
	ironchance[i] = roll(45) - (goldchance[i] + oilchance[i] / 2);
	if (size[i] > 10) {
	    printf("%3d   %3d  %3d  %3d  %3d\n",
		   size[i], oilchance[i], goldchance[i],
		   fertchance[i], ironchance[i]);
	}
    }
    if (i == YbyX)
	printf("Oops.  YbyX not big enough\n");
    sp = sects;
    for (y = 0; y < WORLD_Y; y++) {
	for (x = 0; x < WORLD_X; x++) {
	    if (((x ^ y) & 1))
		continue;
	    makeore(sp++);
	}
    }
    lseek(sectf, 0L, 0);
    write(sectf, sects, sizeof(sects));
    close(sectf);
    exit(0);
}

int cont = 0;

#define max(a,b) (a < b ? b : a)

void
makec(x, y, n)
int x;
int y;
int n;
{
    register int dy;
    register int dx;
    register int dir;
    register short *cp;

    if (s[y][x / 2] == SCT_WATER)
	return;
    cp = &c[y][x / 2];
    if (*cp != 0)
	return;
    if (n == 0)
	n = ++cont;
    *cp = n;
    size[n]++;
    for (dir = 1; dir <= 6; dir++) {
	dx = (diroff[dir][0] + x + WORLD_X) % WORLD_X;
	dy = (diroff[dir][1] + y + WORLD_Y) % WORLD_Y;
	makec(dx, dy, n);
    }
}

void
makeore(sp)
register struct sctstr *sp;
{
    int elev;
    int oil, gmin, iron, fertil, ur;
    int lev;
    int i;
    int x, y;

    y = sp->sct_y;
    x = sp->sct_x;
    elev = e[y][x / 2];
    lev = (elev * 12 + nearelev(x, y)) / 18;
    i = c[y][x / 2];
    oil = 0;
    iron = 0;
    gmin = 0;
    fertil = 0;
    ur = 0;
    if (lev < 20 && lev > -15 && roll(100) < 30 + oilchance[i] + lev) {
	oil = (45 - lev) + roll(50);
	if (oil > 100)
	    oil = 100;
	if (oil < 10)
	    oil = 0;
    }
    if (lev > 15 && roll(20 + lev) < lev + ironchance[i]) {
	iron = lev + roll(55);
	if (iron > 100)
	    iron = 100;
	if (iron < 10)
	    iron = 0;
    }
    if (lev > 16 && roll(35 + lev) < lev + goldchance[i]) {
	gmin = (lev - 5) + roll(45);
	if (gmin > 100)
	    gmin = 100;
	if (gmin < 10)
	    gmin = 0;
    }
    if (lev > 14 && roll(35 + lev) < lev + urchance[i]) {
	ur = (lev - 5) + roll(45);
	if (ur > 100)
	    ur = 100;
	if (ur < 10)
	    ur = 0;
    }
    fertil = (85 + fertchance[i] - lev) - (((iron + gmin) / 2) - oil / 3);
    if (fertil > 120)
	fertil = 120;
    if (fertil < 0)
	fertil = 0;
    sp->sct_oil = 100;
    sp->sct_min = 100;
    sp->sct_gmin = 100;
    sp->sct_fertil = 100;
    sp->sct_uran = 100;
}

int
nearelev(x, y)
int x;
int y;
{
    int dir;
    int dx;
    int dy;
    int level;

    level = 0;
    for (dir = 1; dir <= 6; dir++) {
	dx = (diroff[dir][0] + x + WORLD_X) % WORLD_X;
	dy = (diroff[dir][1] + y + WORLD_Y) % WORLD_Y;
	level += e[dy][dx / 2];
    }
    return level;
}

/*		Already in libgen.a
roll(n)
	int	n;
{
	return random() % n;
}
*/

int
min(a, b)
int a;
int b;
{
    if (a < b)
	return a;
    return b;
}
